src/adapter/src/optimize/materialized_view.rs

```diff
@@ -299,26 +279,44 @@ impl Optimize<GlobalMirPlan<Resolved>> for OptimizeMaterializedView {
         .map_err(OptimizerError::Internal)?;
 
         // Return the plan at the end of this `optimize` step.
-        Ok(GlobalLirPlan { df_desc, df_meta })
+        Ok(GlobalLirPlan {
+            df_desc,
+            df_meta,
+            phantom: PhantomData::<Unresolved>,
+        })
     }
 }
 
-impl GlobalLirPlan {
-    pub fn unapply(self) -> (LirDataflowDescription, DataflowMetainfo) {
-        (self.df_desc, self.df_meta)
-    }
+impl GlobalLirPlan<Unresolved> {
+    /// Produces the [`GlobalLirPlan`] with [`Resolved`] timestamp.
+    pub fn resolve(mut self, as_of: Antichain<Timestamp>) -> GlobalLirPlan<Resolved> {
+        // Set the `as_of` timestamp for the dataflow.
+        self.df_desc.set_as_of(as_of);
 
-    pub fn df_desc(&self) -> &LirDataflowDescription {
-        &self.df_desc
-    }
+        soft_assert_or_log!(
+            self.df_desc.index_exports.is_empty(),
+            "unexpectedly setting until for a DataflowDescription with an index",
+        );
 
-    pub fn df_meta(&self) -> &DataflowMetainfo {
-        &self.df_meta
+        // The only outputs of the dataflow are sinks, so we might be able to
+        // turn off the computation early if they all have non-trivial `up_to`s.
+        self.df_desc.until = Antichain::from_elem(Timestamp::MIN);
+        for (_, sink) in &self.df_desc.sink_exports {
+            self.df_desc.until.join_assign(&sink.up_to);
+        }
```

Comment:
I think we should remove this, as `persist_sink` will crash anyway when it gets an `up_to` that is not the empty frontier. We could replace this with an assert for `sink.up_to.is_empty()`.

Reply:
Done.